contents/java/JUnit.md

```diff
@@ -0,0 +1,234 @@
+
+We can now write
+```java
+assertThrows(ArithmeticException.class, () -> divide(100, 0));
+```
+
+This expression is a lot easier to understand as compared to using plain Java, and as an added bonus, the code is also much shorter to write!
+
+### Before/After hooks
+
+Not every system can be tested so easily. Sometimes, testing is not as straightforward as running a function.
+For example, we might want to store information in a database. We would have to first connect to the database before we can run any test.
+
+JUnit exposes an API that allows developers to do this process easily and clearly. It offers the `@BeforeAll` and `@BeforeEach` tags that allow us to run any code before each test, allowing us to easily initialize our testing environments before we run any tests:
+
+```java
+class TestInvolvingDatabases {
+  Database db;
+  @BeforeAll
+  public static void initializeDatabaseConnection() {
+    // Initialize database connections
+    db = connectToDatabase();
+  }
+  @Test
+  public static void addUserTest() {
+    db.addUser('NewUser');
+    ...
+  }
+}
+```
+
+Similarly, we can use the `@AfterAll` and `@AfterEach` tags to run code after each test:
+
+```java
+class TestInvolvingDatabases {
+  Database db;
+  @AfterAll
+  public static void closeDatabaseConnection() {
+    db.close(); //Explicitly close DB connection instead of timing out.
+  }
+}
+```
+
+The `BeforeEach` API can also be leveraged to ensure that we standardise the testing environment before each test.
+For example, we can clear our database before each test to ensure that tests do not <trigger for="modal:affectEachOther" trigger="click">affect each other</trigger>.
+
+```java
+class TestInvolvingDatabase {
+  Database db;
+  @BeforeEach
+  public static void resetDatabaseTable() {
+    db.truncate("testTable");
+  }
+}
+```
+
+### Extension Model
+
+Consider a project where multiple functions in different files each need to connect to a database with a different URL.
+One way would be to write a separate `initializeDatabaseConnection()` method for each test that requires a database connection.
+
+This would result in having duplicated code, which can introduce errors if we update one method but not another. Writing tests and helper methods once, then reusing these methods across our entire test suite allows us to prevent code duplication errors.
+
+In JUnit, using the extension model allows us to abstract away the idea of connecting to a database before tests. We can now write the database extension once, then reuse this extension across multiple tests.
+
+```java
+public class DatabaseExtension {
+  public static Database connectTo(String url) {
+    Database db = connectToDatabase(url);
+    return db;
+  }
+}
+```
+
+Now, any code requiring a database can simply do
+
+```java
+public class DatabaseTest {
+  @RegisterExtension
+  public static Database db = DatabaseExtension.connectTo('localhost');
+}
+```
+
+to initialize a database connection for all tests.
+
```

Comment:
Why is this better than writing it as a normal static method and calling it wherever it is needed?